exporter/datadogexporter/config/warn_envvars.go

```
@@ -0,0 +1,139 @@
+			HistConfig: HistogramConfig{
+				Mode:         "distributions",
+				SendCountSum: false,
+			},
+		},
+		Traces: TracesConfig{
+			SampleRate:      1,
+			IgnoreResources: []string{},
+		},
+		SendMetadata:        true,
+		UseResourceMetadata: true,
+	}
+}
+
+// errUsedEnvVar constructs an error instructing users not to rely on
+// the automatic environment variable detection and set the environment variable
+// explicitly on configuration instead.
+func errUsedEnvVar(settingName, envVarName string) error {
+	return fmt.Errorf(
+		"%q will not default to %q's value starting on v0.50.0. Set %s: ${%s} to remove this warning",
+		settingName,
+		envVarName,
+		settingName,
+		envVarName,
+	)
+}
+
+// tagsDiffer checks if the host tags from each configuration are different.
+func tagsDiffer(cfgTags []string, futureTags []string) bool {
+	if len(cfgTags) != len(futureTags) {
+		return true
+	}
+
+	oldCfgSet := map[string]struct{}{}
+	for _, tag := range cfgTags {
+		oldCfgSet[tag] = struct{}{}
+	}
+
+	for _, tag := range futureTags {
+		if _, ok := oldCfgSet[tag]; !ok {
+			// tag is in old but not new
+			return true
+		}
+	}
+
+	return false
+}
+
+// warnUseOfEnvVars returns warnings related to automatic environment variable detection.
+// Right now, we automatically get the value for e.g. the API key from DD_API_KEY, even if
+// the user is not doing `api.key: ${DD_API_KEY}`. We are going to remove this functionality.
+func warnUseOfEnvVars(configMap *config.Map, cfg *Config) (warnings []error) {
+	// We don't see the raw YAML contents from our exporter so, to compare with the new default,
+	// we unmarshal the config map on the future default config and compare the two structs.
+	// Any differences will be due to the change in default configuration.
+	futureCfg := futureDefaultConfig()
+	err := configMap.UnmarshalExact(futureCfg)
+	if err != nil {
+		// This should never happen, since unmarshaling passed with the original config.
+		// If it happens, bail out.
```

Isn't there something like `assert` you can use in situations like this one? Silently ignoring that something unexpected occurred seems dangerous.

---

I can pass the error to `Unmarshal`, would you prefer that? That means the exporter will fail to initialize because of this error

---

What about printing the error and then returning?

---

@albertvaka PTAL